dot/parachain/dispute/queues.go

```diff
@@ -0,0 +1,234 @@
+type ParticipationPriority int
+
+const (
+	// ParticipationPriorityBestEffort is the lowest priority
+	ParticipationPriorityBestEffort ParticipationPriority = iota
+	// ParticipationPriorityHigh is the highest priority
+	ParticipationPriorityHigh
+)
+
+// IsPriority returns true if the priority is high
+func (p ParticipationPriority) IsPriority() bool {
+	return p == ParticipationPriorityHigh
+}
+
+var (
+	// errorBestEffortQueueFull is returned when the best effort queue is full and the request could not be processed
+	errorBestEffortQueueFull = errors.New("best effort queue is full")
+	// errorPriorityQueueFull is returned when the priority queue is full and the request could not be processed
+	errorPriorityQueueFull = errors.New("priority queue is full")
+)
+
+// Queue the dispute participation queue
+type Queue interface {
+	// Queue adds a new participation request to the queue
+	Queue(comparator CandidateComparator, request *ParticipationRequest, priority ParticipationPriority) error
+
+	// Dequeue gets the next best request for dispute participation if any.
+	Dequeue() *ParticipationItem
+
+	// PrioritiseIfPresent moves a participation request from the best effort queue to the priority queue
+	PrioritiseIfPresent(comparator CandidateComparator) error
+
+	// PopBestEffort removes the next participation request from the best effort queue
+	PopBestEffort() *ParticipationItem
+
+	// PopPriority removes the next participation request from the priority queue
+	PopPriority() *ParticipationItem
+
+	// Len returns the number of items in the specified queue
+	Len(queueType ParticipationPriority) int
+}
+
+// QueueHandler implements Queue
+// It uses two btree's to store the requests. One for best effort and one for priority.
+// The queues store participationItem's.
+// The btree is ordered by the CandidateComparator of participationItem.
+type QueueHandler struct {
+	bestEffort *btree.BTree
+	priority   *btree.BTree
+
+	bestEffortLock sync.RWMutex
+	priorityLock   sync.RWMutex
+
+	bestEffortMaxSize int
+	priorityMaxSize   int
+
+	//TODO: add metrics
+}
+
+const (
+	// bestEffortQueueSize is the maximum size of the best effort queue
+	bestEffortQueueSize = 100
+	// priorityQueueSize is the maximum size of the priority queue
+	priorityQueueSize = 20000
+)
+
+func (q *QueueHandler) Queue(
+	comparator CandidateComparator,
+	request *ParticipationRequest,
+	priority ParticipationPriority,
+) error {
+	if priority.IsPriority() {
+		if q.Len(ParticipationPriorityHigh) >= q.priorityMaxSize {
+			return errorPriorityQueueFull
+		}
+
+		q.priorityLock.Lock()
+		q.priority.ReplaceOrInsert(newParticipationItem(comparator, request))
+		q.priorityLock.Unlock()
+	} else {
+		if q.Len(ParticipationPriorityBestEffort) >= q.bestEffortMaxSize {
+			return errorBestEffortQueueFull
+		}
+
+		q.bestEffortLock.Lock()
+		q.bestEffort.ReplaceOrInsert(newParticipationItem(comparator, request))
+		q.bestEffortLock.Unlock()
+	}
```

Comment:
I think you're missing some logic here.  Don't you need to remove something from `bestEffort` and then insert into `priority` btree map when `priority.IsPriority() == true`?
```
		if priority.is_priority() {
			if self.priority.len() >= PRIORITY_QUEUE_SIZE {
				return Err(QueueError::PriorityFull)
			}
			// Remove any best effort entry, using it to replace our new
			// request.
			if let Some(older_request) = self.best_effort.remove(&comparator) {
				req.discard_timer();
				req = older_request;
			}
			// Keeping old request if any.
			match self.priority.entry(comparator) {
				Entry::Occupied(_) => req.discard_timer(),
				Entry::Vacant(vac) => {
					vac.insert(req);
				},
			}
			self.metrics.report_priority_queue_size(self.priority.len() as u64);
			self.metrics.report_best_effort_queue_size(self.best_effort.len() as u64);
		} else {
			if self.priority.contains_key(&comparator) {
				// The candidate is already in priority queue - don't
				// add in in best effort too.
				return Ok(())
			}
			if self.best_effort.len() >= BEST_EFFORT_QUEUE_SIZE {
				return Err(QueueError::BestEffortFull)
			}
			// Keeping old request if any.
			match self.best_effort.entry(comparator) {
				Entry::Occupied(_) => req.discard_timer(),
				Entry::Vacant(vac) => {
					vac.insert(req);
				},
			}
			self.metrics.report_best_effort_queue_size(self.best_effort.len() as u64);
		}
```

Reply:
Yes, my bad. Good catch!

Reply:
I've added the missing checks.